market_maker_keeper/api.py

```diff
@@ -0,0 +1,117 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from argparse import Namespace
+
+from market_maker_keeper.band import Bands
+from market_maker_keeper.control_feed import create_control_feed
+from market_maker_keeper.limit import History
+from market_maker_keeper.order_book import OrderBookManager
+from market_maker_keeper.order_history_reporter import create_order_history_reporter
+from market_maker_keeper.price_feed import PriceFeedFactory
+from market_maker_keeper.reloadable_config import ReloadableConfig
+from market_maker_keeper.spread_feed import create_spread_feed
+from market_maker_keeper.util import setup_logging
+
+from pymaker.lifecycle import Lifecycle
+from pymaker.numeric import Wad
+from pyexchange.api import PyexAPI
+
+
+class KeeperAPI:
+    """
+    Define a common abstract API for keepers
+    """
+
+    def __init__(self, arguments: Namespace, pyex_api: PyexAPI):
+
+        setup_logging(arguments)
+
+        self.bands_config = ReloadableConfig(arguments.config)
+        self.price_feed = PriceFeedFactory().create_price_feed(arguments)
+        self.spread_feed = create_spread_feed(arguments)
+        self.control_feed = create_control_feed(arguments)
+        self.order_history_reporter = create_order_history_reporter(arguments)
+
+        self.history = History()
+
+        self.order_book_manager = OrderBookManager(refresh_frequency=arguments.refresh_frequency)
+        self.order_book_manager.get_orders_with(lambda: pyex_api.get_orders(self.pair()))
+        self.order_book_manager.get_balances_with(lambda: pyex_api.get_balances())
+        self.order_book_manager.cancel_orders_with(lambda order: pyex_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders,
+                                                         self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle() as lifecycle:
+            lifecycle.initial_delay(10)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def shutdown(self):
+        self.order_book_manager.cancel_all_orders()
+
+    # Each exchange takes pair input as a different format
+    def pair(self):
+        raise NotImplementedError()
+
+    def token_sell(self) -> str:
+        raise NotImplementedError()
+
+    def token_buy(self) -> str:
+        raise NotImplementedError()
+
+    # Different keys are used to access balance object for different exchanges
+    def our_available_balance(self, our_balances: dict, token: str) -> Wad:
+        raise NotImplementedError()
+
+    def our_sell_orders(self, our_orders: list) -> list:
+        return list(filter(lambda order: order.is_sell, our_orders))
+
+    def our_buy_orders(self, our_orders: list) -> list:
+        return list(filter(lambda order: not order.is_sell, our_orders))
+
+    def synchronize_orders(self):
```

Comment:
I suggest breaking this in syncronize_orders / cancel_orders / place orders methods for reusability

Reply:
Updated accordingly.

Reply:
Almost all of the functionality already seems to be split out? What else would you break out for place or cancel orders - `synchronize_orders` seems to be a pretty barebones control flow that calls those orders. Any which way it would still need those two if checks. That being said, could be not understanding it

Reply:
I am considering that some keepers put different conditions in order to place an order, e.g. you have 

```# Do not place new orders if order book state is not confirmed
        if order_book.orders_being_placed or order_book.orders_being_cancelled:```

while oasis keeper is checking 

```
if order_book.orders_being_placed:
```

https://github.com/makerdao/market-maker-keeper/blob/master/market_maker_keeper/oasis_market_maker_keeper.py#L246

or maybe just create methods to check if order should be canceled / deleted that keepers can override

Reply:
As discussed offline, added a second Abstract class for decentralized exchanges with additional methods.